File: Generator/background_generator.py
import io
from PIL import Image, ImageDraw, ImageFont
import qrcode
import os

# Вспомогательные функции для контраста (WCAG)
def hex_to_rgb(hex_color):
    hex_color = hex_color.lstrip('#')
    return tuple(int(hex_color[i:i+2], 16) for i in (0, 2, 4))

def get_luminance(r, g, b):
    a = [v / 255.0 for v in (r, g, b)]
    a = [v / 12.92 if v <= 0.03928 else ((v + 0.055) / 1.055) ** 2.4 for v in a]
    return a[0] * 0.2126 + a[1] * 0.7152 + a[2] * 0.0722

def get_contrast_ratio(bg_rgb, fg_rgb):
    bg_lum = get_luminance(*bg_rgb)
    fg_lum = get_luminance(*fg_rgb)
    return (max(bg_lum, fg_lum) + 0.05) / (min(bg_lum, fg_lum) + 0.05)

def choose_text_color(bg_hex):
    bg_rgb = hex_to_rgb(bg_hex)
    white_rgb = (255, 255, 255)
    black_rgb = (0, 0, 0)
    white_contrast = get_contrast_ratio(bg_rgb, white_rgb)
    black_contrast = get_contrast_ratio(bg_rgb, black_rgb)
    return '#FFFFFF' if white_contrast >= 4.5 else ('#000000' if black_contrast >= 4.5 else '#FFFFFF')

# Основная функция генератора
def generate_background(employee_data, base_background_path, output_path=None):
    # Загрузка базового фона
    base_img = Image.open(base_background_path)
    base_img = base_img.resize((1920, 1080)).convert('RGBA')
    
    # Overlay градиент на основе corporate_colors (полупрозрачный)
    primary = hex_to_rgb(employee_data['branding']['corporate_colors']['primary'])
    secondary = hex_to_rgb(employee_data['branding']['corporate_colors']['secondary'])
    overlay = Image.new('RGBA', base_img.size, (0, 0, 0, 0))
    overlay_draw = ImageDraw.Draw(overlay)
    for y in range(base_img.height):
        frac = y / base_img.height
        color = tuple(int(primary[i] * (1 - frac) + secondary[i] * frac) for i in range(3)) + (80,)  # alpha ~30%
        overlay_draw.line([(0, y), (base_img.width, y)], fill=color)
    base_img = Image.alpha_composite(base_img, overlay)
    
    # ВАЖНО: Пересоздаем draw после alpha_composite
    draw = ImageDraw.Draw(base_img)
    
    # Добавление логотипа
    logo_path = employee_data['branding']['logo_url']
    try:
        logo = Image.open(logo_path).convert('RGBA')
        logo = logo.resize((200, 200))
        base_img.paste(logo, (50, 50), logo)
    except Exception as e:
        print(f"Warning: Could not load logo from {logo_path}: {e}")

    # Загрузка шрифта с улучшенным fallback
    font_size = 40
    slogan_font_size = 30
    
    # Попробуем разные шрифты
    font_paths = [
        'arial.ttf',
        '/System/Library/Fonts/Arial.ttf',  # Mac
        '/usr/share/fonts/truetype/freefont/FreeSans.ttf',  # Linux
        'C:/Windows/Fonts/arial.ttf',  # Windows
    ]
    
    font = None
    slogan_font = None
    
    for font_path in font_paths:
        try:
            if os.path.exists(font_path):
                font = ImageFont.truetype(font_path, font_size)
                slogan_font = ImageFont.truetype(font_path, slogan_font_size)
                print(f"Using font: {font_path}")
                break
        except Exception:
            continue
    
    # Если шрифты не найдены, используем default
    if font is None:
        print("Warning: No TrueType fonts found. Using default font (text may be small).")
        font = ImageFont.load_default()
        slogan_font = ImageFont.load_default()

    bg_color_for_contrast = employee_data['branding']['corporate_colors']['primary']
    text_color_hex = choose_text_color(bg_color_for_contrast)
    text_color_rgb = hex_to_rgb(text_color_hex)
    shadow_color_rgb = (0, 0, 0) if text_color_hex == '#FFFFFF' else (255, 255, 255)
    
    # Полупрозрачный фон за текстом
    box_color = (0, 0, 0, 128) if text_color_hex == '#FFFFFF' else (255, 255, 255, 128)

    y = 300
    privacy = employee_data['privacy_level']
    
    def draw_text(text, x, y, local_font=font):
        # Получаем размеры текста
        try:
            # Для PIL >= 8.0.0 используем textbbox
            if hasattr(draw, 'textbbox'):
                text_bbox = draw.textbbox((x, y), text, font=local_font)
            else:
                # Для старых версий используем textsize
                text_width, text_height = draw.textsize(text, font=local_font)
                text_bbox = (x, y, x + text_width, y + text_height)
        except Exception as e:
            print(f"Error calculating text size: {e}")
            # Fallback bbox
            text_bbox = (x, y, x + len(text) * 20, y + 40)
        
        box_padding = 10
        draw.rectangle(
            (text_bbox[0] - box_padding, text_bbox[1] - box_padding, 
             text_bbox[2] + box_padding, text_bbox[3] + box_padding),
            fill=box_color
        )
        # Тень и текст
        draw.text((x + 2, y + 2), text, font=local_font, fill=shadow_color_rgb)
        draw.text((x, y), text, font=local_font, fill=text_color_rgb)
        return text_bbox[3] - text_bbox[1]  # возвращаем высоту текста

    # Рисуем текст с информацией
    try:
        text_height = draw_text(employee_data['full_name'], 50, y)
        y += text_height + 20
        
        text_height = draw_text(employee_data['position'], 50, y)
        y += text_height + 20
        
        if privacy in ['medium', 'high']:
            text_height = draw_text(employee_data['company'], 50, y)
            y += text_height + 20
            
            text_height = draw_text(employee_data['department'], 50, y)
            y += text_height + 20
            
            text_height = draw_text(employee_data['office_location'], 50, y)
            y += text_height + 20
        
        if privacy == 'high':
            text_height = draw_text(employee_data['contact']['email'], 50, y)
            y += text_height + 20
            
            text_height = draw_text(employee_data['contact']['telegram'], 50, y)
            y += text_height + 20
            
            # QR для email
            qr = qrcode.QRCode(box_size=5)
            qr.add_data(f"mailto:{employee_data['contact']['email']}")
            qr.make(fit=True)
            qr_img = qr.make_image(fill_color="black", back_color="white").resize((128, 128))
            # Конвертируем в RGBA для прозрачности
            qr_img = qr_img.convert('RGBA')
            base_img.paste(qr_img, (50, y))
            y += 150
            
            # QR для telegram
            tg_username = employee_data['contact']['telegram'].replace('@', '')
            qr = qrcode.QRCode(box_size=5)
            qr.add_data(f"https://t.me/{tg_username}")
            qr.make(fit=True)
            qr_img = qr.make_image(fill_color="black", back_color="white").resize((128, 128))
            qr_img = qr_img.convert('RGBA')
            base_img.paste(qr_img, (50, y))
            y += 150
        
        # Slogan внизу
        slogan_height = draw_text(employee_data['branding']['slogan'], 50, base_img.height - 80, local_font=slogan_font)
        
    except Exception as e:
        print(f"Error drawing text: {e}")

    # Сохранение
    if output_path:
        base_img.save(output_path, format='PNG')
        print(f"Background saved to: {output_path}")
        return output_path
    else:
        buffered = io.BytesIO()
        base_img.save(buffered, format='PNG')
        buffered.seek(0)
        return buffered.read()

# Пример использования
if __name__ == "__main__":
    example_data = {
        "full_name": "Иванов Сергей Петрович",
        "position": "Ведущий инженер по компьютерному зрению",
        "company": "ООО «Рога и Копыта»",
        "department": "Департамент компьютерного зрения",
        "office_location": "Новосибирск, технопарк «Идея»",
        "contact": {
            "email": "sergey.ivanov@t1dp.ru",
            "telegram": "@sergey_vision"
        },
        "branding": {
            "logo_url": "./testData/logos/medium_9eb9cd58b9ea5e04c890326b5c1f471f.png",
            "corporate_colors": {
                "primary": "#0052CC",
                "secondary": "#00B8D9"
            },
            "slogan": "Инновации в каждый кадр"
        },
        "privacy_level": "high"
    }
    
    base_path = "./testData/Back/white.png"
    output = generate_background(example_data, base_path, output_path="generated_background.png")
    print(f"Generated: {output}")

File: Generator/background_generator.js
// === ВСПОМОГАТЕЛЬНЫЕ ФУНКЦИИ ===
function hexToRgb(hex) {
    hex = hex.replace('#', '');
    if (hex.length === 3) {
        hex = hex.split('').map(c => c + c).join('');
    }
    return {
        r: parseInt(hex.substr(0, 2), 16),
        g: parseInt(hex.substr(2, 2), 16),
        b: parseInt(hex.substr(4, 2), 16)
    };
}

function getLuminance(r, g, b) {
    const a = [r, g, b].map(v => {
        v /= 255;
        return v <= 0.03928 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4);
    });
    return a[0] * 0.2126 + a[1] * 0.7152 + a[2] * 0.0722;
}

function getContrastRatio(bg, fg) {
    const bgLum = getLuminance(bg.r, bg.g, bg.b);
    const fgLum = getLuminance(fg.r, fg.g, fg.b);
    return (Math.max(bgLum, fgLum) + 0.05) / (Math.min(bgLum, fgLum) + 0.05);
}

function chooseTextColor(bgHex) {
    const bg = hexToRgb(bgHex);
    const white = { r: 255, g: 255, b: 255 };
    const black = { r: 0, g: 0, b: 0 };
    const whiteContrast = getContrastRatio(bg, white);
    const blackContrast = getContrastRatio(bg, black);
    return whiteContrast >= 4.5 ? '#FFFFFF' : (blackContrast >= 4.5 ? '#000000' : '#FFFFFF');
}

// === ГЕНЕРАЦИЯ ИЗОБРАЖЕНИЯ ===
async function generateBackground(baseBackgroundPath) {
    // Получаем данные сотрудника из глобальной переменной
    const employee = window.employeeInstance;
    if (!employee) {
        console.error('window.employeeInstance не определён');
        return null;
    }
    const info = employee.employee;
    
    // Создаём canvas
    const WIDTH = 1920, HEIGHT = 1080;
    const canvas = document.createElement('canvas');
    canvas.width = WIDTH;
    canvas.height = HEIGHT;
    const ctx = canvas.getContext('2d');

    // Загрузка фона
    try {
        const baseImg = new Image();
        baseImg.crossOrigin = 'anonymous';
        await new Promise((resolve, reject) => {
            baseImg.onload = resolve;
            baseImg.onerror = reject;
            baseImg.src = baseBackgroundPath;
        });
        ctx.drawImage(baseImg, 0, 0, WIDTH, HEIGHT);
    } catch (err) {
        console.warn('Не удалось загрузить фоновое изображение:', err.message);
        return null;
    }

    // Градиентный оверлей (всегда)
    const primary = hexToRgb(info.branding.corporate_colors.primary);
    const secondary = hexToRgb(info.branding.corporate_colors.secondary);
    for (let y = 0; y < HEIGHT; y++) {
        const frac = y / HEIGHT;
        const r = Math.round(primary.r * (1 - frac) + secondary.r * frac);
        const g = Math.round(primary.g * (1 - frac) + secondary.g * frac);
        const b = Math.round(primary.b * (1 - frac) + secondary.b * frac);
        ctx.fillStyle = `rgba(${r}, ${g}, ${b}, 0.12)`;
        ctx.fillRect(0, y, WIDTH, 1);
    }

    // Логотип (всегда, если URL)
    if (info.branding.logo_url) {
        try {
            const logo = new Image();
            logo.crossOrigin = 'anonymous';
            await new Promise((resolve, reject) => {
                logo.onload = resolve;
                logo.onerror = reject;
                logo.src = info.branding.logo_url;
            });
            const logoSize = 125;
            ctx.drawImage(logo, 50, 50, logoSize, logoSize);
        } catch (err) {
            console.warn('Не удалось загрузить логотип:', err.message);
        }
    }

    // Шрифты
    const fontFamily = 'Arial';
    const font40 = `40px ${fontFamily}`;
    const font30 = `30px ${fontFamily}`;
    const font25 = `25px ${fontFamily}`;
    ctx.font = font40;

    // Цвета
    const primaryColor = info.branding.corporate_colors.primary;
    const textColor = chooseTextColor(primaryColor);
    const textColorRgb = hexToRgb(textColor);
    const shadowColor = textColorRgb.r > 128 ? 'rgba(0,0,0,0.5)' : 'rgba(255,255,255,0.5)';
    const primaryHex = primaryColor.replace('#', '');
    const secondaryHex = info.branding.corporate_colors.secondary.replace('#', '');
    const boxColors = {
        primary: `#${primaryHex}77`,
        secondary: `#${secondaryHex}64`
    };

    // Вспомогательные функции (с исправлением для пустых текстов)
    function setFont(size) {
        ctx.font = `${size}px ${fontFamily}`;
    }

    function getTextWidth(text, size = 40) {
        setFont(size);
        return ctx.measureText(text || '').width;
    }

    function splitText(text, maxWidth, fontSize = 40) {
        if (!text) return [];
        setFont(fontSize);
        const words = text.split(' ');
        const lines = [];
        let current = [];
        for (const word of words) {
            const test = [...current, word].join(' ');
            if (ctx.measureText(test).width <= maxWidth) {
                current.push(word);
            } else {
                if (current.length) lines.push(current.join(' '));
                current = [word];
            }
        }
        if (current.length) lines.push(current.join(' '));
        return lines;
    }

    function drawRoundedRect(x, y, w, h, r, fill) {
        ctx.beginPath();
        ctx.moveTo(x + r, y);
        ctx.lineTo(x + w - r, y);
        ctx.quadraticCurveTo(x + w, y, x + w, y + r);
        ctx.lineTo(x + w, y + h - r);
        ctx.quadraticCurveTo(x + w, y + h, x + w - r, y + h);
        ctx.lineTo(x + r, y + h);
        ctx.quadraticCurveTo(x, y + h, x, y + h - r);
        ctx.lineTo(x, y + r);
        ctx.quadraticCurveTo(x, y, x + r, y);
        ctx.closePath();
        if (fill) {
            ctx.fillStyle = fill;
            ctx.fill();
        }
    }

    function drawTextWithBox(text, x, y, options = {}) {
        const {
            fontSize = 40,
            boxType = 'primary',
            align = 'left',
            vPad = 12,
            lineSpacing = 5
        } = options;
        setFont(fontSize);
        const lines = Array.isArray(text) ? text : splitText(text, 500, fontSize);
        if (lines.length === 0) return 0; // Пропуск пустого текста
        const lineHeights = lines.map(() => fontSize * 0.8);
        const totalHeight = lineHeights.reduce((a, b) => a + b, 0) + lineSpacing * (lines.length - 1);
        const maxWidth = Math.max(...lines.map(line => ctx.measureText(line).width), 1); // Минимум 1 для избежания ошибок
        let startX = x;
        if (align === 'right') startX = WIDTH - maxWidth - 70;
        if (align === 'center') startX = (WIDTH - maxWidth) / 2;
        const boxLeft = startX - 10;
        const boxRight = startX + maxWidth + 10;
        const boxTop = y;
        const boxBottom = y + totalHeight + 2 * vPad;
        drawRoundedRect(boxLeft, boxTop, boxRight - boxLeft, boxBottom - boxTop, 15, boxColors[boxType]);
        let currentY = y + vPad;
        lines.forEach((line, i) => {
            const lineWidth = ctx.measureText(line).width;
            const lineX = align === 'right' ? WIDTH - lineWidth - 70 :
                         align === 'center' ? (WIDTH - lineWidth) / 2 : startX;
            ctx.fillStyle = shadowColor;
            ctx.fillText(line, lineX + 2, currentY + 2);
            ctx.fillStyle = textColor;
            ctx.fillText(line, lineX, currentY);
            currentY += lineHeights[i] + lineSpacing;
        });
        return totalHeight + 2 * vPad;
    }

    // === ОСНОВНАЯ ЛОГИКА РАСПОЛОЖЕНИЯ ===
    // Low и выше: имя, должность
    const positionLines = splitText(info.position, 500, 40);
    const leftLines = [...positionLines, info.full_name].filter(line => line);
    if (leftLines.length > 0) {
        const leftBlockHeight = drawTextWithBox(leftLines, 50, HEIGHT - 200, {
            vPad: 12,
            lineSpacing: 5
        });
        const leftStartY = HEIGHT - leftBlockHeight - 50;
        drawTextWithBox(leftLines, 50, leftStartY, { vPad: 12, lineSpacing: 5 });
    }

    // Medium и выше: компания, отдел, локация
    if (['medium', 'high'].includes(info.privacy_level)) {
        const rightLines = [info.company, info.department, info.office_location].filter(Boolean);
        if (rightLines.length > 0) {
            drawTextWithBox(rightLines, 0, 120, {
                boxType: 'secondary',
                align: 'right',
                vPad: 12,
                lineSpacing: 5
            });
        }
    }

    // High: контакты (QR)
    if (info.privacy_level === 'high') {
        const qrSize = 128;
        const qrPadding = 20;
        const totalQrWidth = qrSize * 2 + qrPadding;
        const startX = WIDTH - totalQrWidth - 100;
        const yBr = HEIGHT - 220;
        drawTextWithBox("Контакты", 0, yBr - 50, { align: 'center', vPad: 12 });
        const qrData = [];
        if (info.contact.email) qrData.push([`mailto:${info.contact.email}`, "Email", 'primary']);
        if (info.contact.telegram) qrData.push([`https://t.me/${info.contact.telegram.replace('@', '')}`, "Telegram", 'secondary']);
        for (let i = 0; i < qrData.length; i++) {
            const [data, label, boxType] = qrData[i];
            const xPos = startX + i * (qrSize + qrPadding);
            drawRoundedRect(xPos, yBr, qrSize + 20, qrSize + 20, 15, boxColors[boxType]);
            const qrCanvas = document.createElement('canvas');
            qrCanvas.width = qrSize;
            qrCanvas.height = qrSize;
            try {
                await QRCode.toCanvas(qrCanvas, data, {
                    width: qrSize,
                    color: { dark: '#000000', light: '#FFFFFF' }
                });
                ctx.drawImage(qrCanvas, xPos + 10, yBr + 10);
            } catch (err) {
                console.warn('Не удалось сгенерировать QR-код:', err.message);
            }
            drawTextWithBox(label, xPos + 10, yBr + qrSize + 30, {
                fontSize: 25,
                boxType,
                align: 'center',
                vPad: 8
            });
        }
    }

    // Слоган (всегда, если есть)
    if (info.branding.slogan) {
        drawTextWithBox(info.branding.slogan, 0, 50, {
            fontSize: 30,
            boxType: 'secondary',
            align: 'center',
            vPad: 12
        });
    }

    // Возвращаем dataURL
    return canvas.toDataURL('image/png');
}